zkml/src/layers/transformer/positional.rs

```diff
@@ -82,3 +225,441 @@ impl<N: Number> OpInfo for Positional<N> {
+        ensure!(
+            input_scaling.len() == 1,
+            "Expected 1 input scaling factor for positional layer, found {}",
+            input_scaling.len()
+        );
+
+        let Positional::Learned(pos) = self else {
+            unimplemented!("Quantization not implemented for Positional::Rope")
+        };
+
+        // quantize positional matrix
+        let max = pos.positional_matrix.max_abs_output();
+        let pos_scaling = ScalingFactor::from_absolute_max(max, None);
+
+        let quantized_add =
+            pos.add_layer
+                .quantize_op::<S>(data, node_id, &[input_scaling[0], pos_scaling])?;
+
+        let quantized_pos = LearnedPositional {
+            positional_matrix: pos.positional_matrix.quantize(&pos_scaling),
+            unpadded_shape: pos.unpadded_shape,
+            add_layer: quantized_add.quantized_op,
+        };
+
+        Ok(QuantizeOutput {
+            quantized_op: Positional::Learned(quantized_pos),
+            output_scalings: quantized_add.output_scalings,
+            requant_layer: quantized_add.requant_layer,
+        })
+    }
+}
+
+impl PadOp for Positional<Element> {
+    fn pad_node(mut self, _si: &mut ShapeInfo) -> anyhow::Result<Self>
+    where
+        Self: Sized,
+    {
+        match &mut self {
+            Positional::Learned(pos) => {
+                pos.positional_matrix = pos.positional_matrix.pad_next_power_of_two()
+            }
+            Positional::Rope => (),
+        }
+
+        // no need to change `si` since the layer doesn't change the input shapes
+
+        Ok(self)
+    }
+}
+
+impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ProvableOp<E, PCS>
+    for Positional<Element>
+{
+    type Ctx = PositionalCtx;
+
+    fn prove<T: Transcript<E>>(
+        &self,
+        node_id: NodeId,
+        _ctx: &Self::Ctx,
+        last_claims: Vec<&Claim<E>>,
+        step_data: &StepData<E, E>,
+        prover: &mut Prover<E, T, PCS>,
+    ) -> anyhow::Result<Vec<Claim<E>>> {
+        ensure!(
+            last_claims.len() == step_data.inputs.len(),
+            "Found different number of inputs and outputs when proving positional layer: {} inputs, {} outputs",
+            step_data.inputs.len(),
+            last_claims.len(),
+        );
+        let Self::Learned(pos) = self else {
+            unimplemented!("Proving not implemented for Positional::Rope")
+        };
+
+        let mut output_claims = vec![];
+        let mut common_claims = HashMap::new();
+        let proofs = last_claims
+            .into_iter()
+            .zip(&step_data.inputs)
+            .map(|(output_claim, input)| {
+                // derive sub-matrix to be added to input. ToDo: place it in proving data
+                let matrix_slice = TensorSlice::from(&pos.positional_matrix);
+                let sub_pos = matrix_slice
+                    .slice_over_first_dim(0, input.get_shape()[0])
+                    .to_fields();
```

Comment:
Was it not possible to use `Tensor::slice_on_dim()` ?

Reply:
Because `slice_over_first_dim` is returning a subset of the rows of the tensor, while `slice_on_dim` is basically creating a single row afaiu

Reply:
But I would need to go each time through the iterator, collect the rows that I want and build the shape of a new tensor. Looks better to me to have this method that already does this on a slice of the tensor. Wdyt?